go/pkg/services/logcollector/server.go

```diff
@@ -0,0 +1,447 @@
+		return &protologcollector.GetLogsResponse{
+			Success: false,
+			Error:   err.Error(),
+		}, err
+	}
+
+	return &protologcollector.GetLogsResponse{
+		Success: true,
+		Logs:    buffer,
+	}, nil
+}
+
+// startLogStreaming streams logs from a pod and writes them into a file
+func (s *LogCollectorServer) startLogStreaming(ctx context.Context, runId, podName string) {
+
+	// create a log file to the pod
+	logFilePath := s.resolvePodLogFilePath(runId, podName)
+	if err := common.EnsureFileExists(logFilePath); err != nil {
+		s.Logger.ErrorWithCtx(ctx,
+			"Failed to ensure log file",
+			"runId", runId,
+			"logFilePath", logFilePath)
+		return
+	}
+
+	// get logs from pod, and keep the stream open (follow)
+	podLogOptions := &v1.PodLogOptions{
+		Follow: true,
+	}
+	restClientRequest := s.kubeClientSet.CoreV1().Pods(s.namespace).GetLogs(podName, podLogOptions)
+
+	// initialize stream and error for the while loop
+	var (
+		streamErr error
+		stream    io.ReadCloser
+	)
+
+	// stream logs - retry if failed
+	for {
+		stream, streamErr = restClientRequest.Stream(ctx)
+		if streamErr == nil {
+			break
+		}
+		s.Logger.WarnWithCtx(ctx, "Failed to get pod log read/closer", "runId", runId)
+		time.Sleep(1 * time.Second)
+	}
+	defer stream.Close() // nolint: errcheck
+
+	for {
+		keepLogging, err := s.streamPodLogs(ctx, runId, logFilePath, stream)
+		if err != nil {
+			s.Logger.WarnWithCtx(ctx, "An error occurred while streaming pod logs", "err", err)
+		}
+		if keepLogging {
+			continue
+		}
+		break
+	}
+
+	s.Logger.DebugWithCtx(ctx,
+		"Removing item from state file",
+		"runId", runId,
+		"podName", podName)
+
+	// remove run from state file
+	if err := s.stateStore.RemoveLogItem(runId); err != nil {
+		s.Logger.WarnWithCtx(ctx, "Failed to remove log item from state file")
+	}
+}
+
+func (s *LogCollectorServer) streamPodLogs(ctx context.Context,
+	runId,
+	logFilePath string,
+	stream io.ReadCloser) (bool, error) {
+
+	// create a reader from the stream
+	streamReader := bufio.NewReader(stream)
+
+	// peek into the stream, and wait until there is something to read from it
+	for {
+		peekBuf, err := streamReader.Peek(1)
+		if err != nil {
+			s.Logger.WarnWithCtx(ctx, "Failed to peek into stream", "runId", runId)
+		}
+		if len(peekBuf) > 0 {
+			break
+		}
+		time.Sleep(s.readLogWaitTime)
+	}
+
+	// open log file
+	openFlags := os.O_RDWR | os.O_APPEND
+	file, err := os.OpenFile(logFilePath, openFlags, 0644)
+	if err != nil {
+		s.Logger.WarnWithCtx(ctx, "Failed to open file", "err", err, "logFilePath", logFilePath)
+		return true, err
+	}
+	defer file.Close() // nolint: errcheck
+
+	// get a buffer from the pool
+	buf := s.bufferPool.Get().([]byte)
+	defer s.bufferPool.Put(&buf)
+
+	// copy the stream into the file using the buffer.
+	// this is blocking until there is something to read / buffer is full.
+	numBytesWritten, err := io.CopyBuffer(file, streamReader, buf)
```

Comment:
i think to make sure that this is not blokcing while `ctx` is dead, you need to open go routine and close the reader/writer or so so copy buffer will raise an err and ctx would cancel that function